src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -0,0 +1,749 @@
+            self._column_states = initiate_column_states(
+                [attr.name for attr in self.snowflake_plan.attributes]
+            )
+        return self._column_states
+
+
+class SelectableEntity(Selectable):
+    """Query from a table, view, or any other Snowflake objects.
+    Mainly used by session.table().
+    """
+
+    def __init__(self, entity_name, *, analyzer=None) -> None:
+        super().__init__(analyzer)
+        self.entity_name = entity_name
+
+    def sql_query(self) -> str:
+        return self.entity_name
+
+    def schema_query(self) -> str:
+        return f"{analyzer_utils.SELECT} * {analyzer_utils.FROM} {self.entity_name}"
+
+
+class SelectSQL(Selectable):
+    """Query from a SQL. Mainly used by session.sql()"""
+
+    def __init__(self, sql: str, *, analyzer=None, to_select: bool = False) -> None:
+        super().__init__(analyzer)
+        self.to_select = to_select
+        self.original_sql = sql
+        self._schema_query = sql
+        is_select = sql.strip().lower().startswith("select")
+        if to_select and not is_select:
+            self.pre_actions = [Query(sql)]
+            self.sql = result_scan_statement(self.pre_actions[0].query_id_place_holder)
+            self._schema_query = analyzer_utils.schema_value_statement(
+                analyze_attributes(sql, self.analyzer.session)
+            )  # Change to subqueryable schema query so downstream query plan can describe the SQL
+        else:
+            self.sql = sql
+
+    def sql_query(self) -> str:
+        return self.sql
+
+    def schema_query(self) -> str:
+        return self._schema_query
+
+    def to_subqueryable(self):
+        if self.to_select:
+            return self
+        new = SelectSQL(self.sql, to_select=True, analyzer=self.analyzer)
+        new._column_states = self.column_states
+        return new
+
+
+class SelectSnowflakePlan(Selectable):
+    """Wrap a SnowflakePlan to a subclass of Selectable."""
+
+    def __init__(self, snowflake_plan: LogicalPlan, *, analyzer=None) -> None:
+        super().__init__(analyzer)
+        self._snowflake_plan = (
+            snowflake_plan
+            if isinstance(snowflake_plan, SnowflakePlan)
+            else analyzer.resolve(snowflake_plan)
+        )
+        self.pre_actions = self._snowflake_plan.queries[:-1]
+        self.post_actions = self._snowflake_plan.post_actions
+
+    @property
+    def snowflake_plan(self):
+        return self._snowflake_plan
+
+    def sql_query(self) -> str:
+        return self._snowflake_plan.queries[-1].sql
+
+    def schema_query(self) -> str:
+        return self.snowflake_plan.schema_query
+
+
+class SelectStatement(Selectable):
+    """The main logic plan to be used by a DataFrame.
+    It structurally has the parts of a query and uses the ColumnState to decide whether a query can be flattened."""
+
+    def __init__(
+        self,
+        *,
+        projection_: Optional[List[Union[Expression]]] = None,
```

Comment:
```suggestion
        projection_: Optional[List[Expression]] = None,
```

do we need a Union here?